3.8/graphs-custom-pregel.md

```diff
@@ -0,0 +1,1511 @@
+      - `array of strings`: Represents a nested path
+
+- **vertexAccumulators** (object, _optional_):
+  Definition of all used [vertex accumulators](#vertex-accumulators).
+  <!-- TODO: Vertex Accumulators -->
+
+- **globalAccumulators** (object, _optional_):
+  Definition all used global accumulators.
+  [Global Accumulators](#global-accumulator) are able to access variables at shared global level.
+
+- **customAccumulators** (object, _optional_):
+  Definition of all used [custom accumulators](#custom-accumulator).
+
+- **phases** (array, _optional?_):
+  Array of a single or multiple [phase definitions](#phases).
+
+- **debug** (optional): See [Debugging](#debugging).
+
+Phases
+------
+
+Phases will run sequentially during your Pregel computation. The definition of
+multiple phases is allowed. Each phase requires instructions based on the
+operations you want to perform. The initialization program (1) will run a
+single time in the very first round. All upcoming rounds will execute the
+update program (2).
+
+The Pregel program execution will follow the order:
+
+Step 1: Initialization
+1. `onPreStep` (Conductor, executed on Coordinator instances)
+2. `initProgram` (Worker, executed on DB-Server instances)
+3. `onPostStep` (Conductor)
+
+Step 2 (+n): Computation
+1. `onPreStep` (Conductor)
+2. `updateProgram` (Worker)
+3. `onPostStep` (Conductor)
+
+#### Phase parameters
+
+- **name** (string, _required_): Phase name
+  The given name of the defined phase.
+- **onPreStep**: Program as `array of operations` to be executed.
+  The _onPreStep_ program will run **once before** each Pregel execution round.
+- **initProgram**: Program as `array of operations` to be executed.
+  The init program will run **initially once** per all vertices that are part
+  of your graph.
+- **updateProgram**: Program as `array of operations` to be executed.
+  The _updateProgram_ will be executed **during every Pregel execution round**
+  and each **per vertex**.
+- **onPostStep**: Program as `array of operations` to be executed.
+  The _onPostStep_ program will run **once after** each Pregel execution round.
+
+`initProgram` and `updateProgram` return value is inspected. If it is not
+`none`, it must be one of the following:
+- `"vote-halt"` or `false`:
+  indicates that this vertex voted halt.
+- `"vote-active"` or `true`:
+  indicates that this vertex voted active and is active in the next round.
+
+Debugging
+---------
+
+Using the `debug` field in the algorithm specification you instruct the Pregel
+system to generate additional tracing information for debugging purpose.
+Currently, only sent messages can be traced but in future this will be expanded
+as needed.
+
+```js
+{
+  debug: {
+    traceMessages: {
+      "my-vertex-id": {}
+    }
+  }
+}
+```
+
+This will generate a _report_ for every message that is send to the vertex
+`my-vertex-id`. Additionally you can specify a filter by adding a `filter` field.
```

Comment:
> This will generate a _report_ for every message that is sent to the vertex `my-vertex-id`.

I think it is clear, that there will be a report for every message that has `my-vertex-id` as receiver. 
Maybe you wanted to know if we trace _messages sent **to**_ or messages _received by_, i.e. does the sender generate the report or the receiving vertex?

Reply:
I think my confusion came from

> Currently, only sent messages can be traced but in future this will be expanded as needed.

which I interpreted wrongly in the context. So the wording is fine.

Though I think it (i.e. we can only specify a receiver to trace) might be unintuitive at this point. It might be helpful to emphasize that, and add a comment as to why the other way round is not possible/needed.

Reply:
It is possible and needed. But we decided not to add it for the experimental state.